docs/PLUGIN.md

```diff
@@ -0,0 +1,314 @@
+# Custom Plugins
+
+Custom plugins are a powerful way to extend GPUd's monitoring capabilities. They allow you to create custom health checks and monitoring scripts that integrate seamlessly with GPUd's component system.
+
+## Overview
+
+Plugins in GPUd serve two main purposes:
+
+1. **Initialization (init)**: One-time setup tasks that run when GPUd starts
+2. **Monitoring (component)**: Regular health checks that run periodically
+
+Plugins can be configured to run as:
+- Single components (one health check)
+- Multiple components (multiple health checks from a single configuration)
+- Initialization tasks (one-time setup)
+
+## Plugin Types
+
+### Init Plugins
+Init plugins are used for one-time setup tasks when GPUd starts. They're perfect for:
+- Setting up system configurations
+- Installing required dependencies
+- Initializing monitoring tools
+- Setting up logging
+
+### Component Plugins
+Component plugins run periodically to monitor system health. They can be:
+- **Single Component**: One health check with one configuration
+- **Multi-Component**: Multiple health checks from a single configuration, either specified directly or loaded from a file
+
+## Configuration Format
+
+Plugins are configured using YAML files. Here's the basic structure:
+
+```yaml
+plugin_name: string  # Required, max 128 chars
+type: string        # Required, one of: "init", "component", "component_list"
+run_mode: string    # Optional, defaults to "manual"
```

Comment:
```suggestion
run_mode: string    # Optional, defaults to "auto"
```

Do we want to treat empty `run_mode` as manual? The current logic is, we only check `manual` to skip periodic check calls. All other values including empty are NOT being used.

https://github.com/leptonai/gpud/blob/4cbd7434e79d26aade05d127562c9d425cb49cce/pkg/custom-plugins/component.go#L69-L72